README.md
# CryptoParser

## What is it and what is it not?

As the project name CryptoParser implies, it is a cryptographic protocol parser. The main purpose of creating this
library is the fact, that cryptography protocol analysis differs in many aspect from establishing a connection using a
cryptographic protocol. Analysis is mostly testing where we trigger special and corner cases of the protocol and we also
trying to establish connection with hardly supported, experimental, obsoleted or even deprecated mechanisms or
algorithms which are may or may not supported by the latest or any version of an implementation of the cryptographic
protocol.

As follows, it is neither a comprehensive nor a secure implementation of any cryptographic protocol. On the one hand
library implements only the absolutely necessary parts of the protocol. On the other it contains completely insecure
algorithms and mechanisms. It is not designed and contraindicated to use this library establishing secure connections.
If you are searching for cryptographic protocol implementation, there are several existing wrappers and native
implementations for Python (eg: M2Crypto, pyOpenSSL, Paramiko, \...).

## Quick start

CryptoParser can be installed directly via pip:

```shell
$ pip install cryptoparser
```

## Development environment

If you want to setup a development environment, you are in need of [pipenv](https://docs.pipenv.org/).

```shell
$ cd cryptoparser
$ pipenv install --dev
$ pipenv shell
```

## Features

### Protocols

- Secure Shell (SSH)
  - [SSH 2.0](https://tools.ietf.org/html/rfc4253)
- Secure Socket Layer (SSL)
  - [SSL 2.0](https://tools.ietf.org/html/draft-hickman-netscape-ssl-00)
  - [SSL 3.0](https://tools.ietf.org/html/rfc6101)
- Transport Layer Security (TLS)
  - [TLS 1.0](https://tools.ietf.org/html/rfc2246)
  - [TLS 1.1](https://tools.ietf.org/html/rfc4346)
  - [TLS 1.2](https://tools.ietf.org/html/rfc5246)
  - [TLS 1.3](https://tools.ietf.org/html/rfc8446)

### Python implementation

- CPython (2.7, \>=3.3)
- PyPy (2.7, 3.5)

### Operating systems

- Linux
- macOS
- Windows

## Protocol Specific Features

### Hypertext Transfer Protocol (HTTP)

#### Headers

1.  supports header wire format parsing
2.  supports detailed parsing of generic headers
    ([Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type),
    [Server](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server))
3.  supports detailed parsing of caching headers ([Age](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Age),
    [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control),
    [Date](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Date),
    [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag),
    [Expires](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expires),
    [Last-Modified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified),
    [Pragma](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Pragma))
4.  supports detailed parsing of security headers
    ([Expect-CT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expect-CT),
    [Expect-Staple](https://scotthelme.co.uk/designing-a-new-security-header-expect-staple),
    [Referrer-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy),
    [Strict-Transport-Security](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security),
    [X-Content-Type-Options](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options),
    [X-Frame-Options](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options))

### Transport Layer Security (TLS)

Only features that cannot be or difficultly implemented by some of the most popular SSL/TLS implementations (eg:
[GnuTls](https://www.gnutls.org/), [LibreSSL](https://www.libressl.org/), [OpenSSL](https://www.openssl.org/),
[wolfSSL](https://www.wolfssl.com/), \...) are listed.

#### Generic

1.  supports [Generate Random Extensions And Sustain
    Extensibility](https://tools.ietf.org/html/draft-ietf-tls-grease-04) (GREASE) values for
    - protocol version
    - extension type
    - ciphers suite
    - signature algorithms
    - named group
2.  supports easy [JA3
    fingerprint](https://engineering.salesforce.com/tls-fingerprinting-with-ja3-and-ja3s-247362855967) generation

#### Protocol Versions

1.  support not only the final, but also draft versions

#### Cipher Suites

1.  supports each cipher suites discussed on [ciphersuite.info](https://ciphersuite.info)
2.  supports [GOST](https://en.wikipedia.org/wiki/GOST) (national standards of the Russian Federation and CIS countries)
    cipher suites

### Secure Shell (SSH)

#### Cipher Suites

1.  identifies as much encryption algorithms as possible (more than 200, compared to 70+ currently supported by OpenSSH)
2.  supports [HASSH fingerprint](https://engineering.salesforce.com/open-sourcing-hassh-abed3ae5044c/) calculation
    (\#96)

#### Public Keys

1.  supports host keys and cretificates (both `V00` and `V01`)

## License

The code is available under the terms of Mozilla Public License Version 2.0 (MPL 2.0).

A non-comprehensive, but straightforward description of MPL 2.0 can be found at [Choose an open source
license](https://choosealicense.com/licenses#mpl-2.0) website.


CHANGELOG.md
# Changelog

## 0.8.4 - 2023-01-22

- TLS (`tls`)
  - Generic
    - protocol item classes for MySQL support (\#61)

## 0.8.2 - 2022-10-10

- TLS (`tls`)
  - Cipher Suites (`ciphers`)
    - add OpenSSL names (\#54)
    - add min/max versions (\#55)
- SSH (`ssh`)
  - Public Keys (`pubkeys`)
    - [HASSH fingerprint](https://engineering.salesforce.com/open-sourcing-hassh-abed3ae5044c/) calculation (\#48)
    - add [host certificate](https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.certkeys) related classes
      (\#53)

## 0.8.0 - 2022-01-18

- SSH (`ssh`)
  - Public Keys (`pubkeys`)
    - add [public key](https://datatracker.ietf.org/doc/html/rfc4253#section-6.6) related classes (\#43)
  - Versions (`versions`)
    - add [software version](https://tools.ietf.org/html/rfc4253#section-4.2) related classes (\#46)

## 0.7.3 - 2021-12-26

- Generic
  - Fix time zone handlind in datetime parser

## 0.7.2 - 2021-10-07

### Other

- switch to Markdown format in changelog, readme and contributing
- update contributing to the latest version from contribution-guide.org

## 0.7.1 - 2021-09-20

- TLS (`tls`)
  - protocol item classes for RDP support (\#44)

## 0.7.0 - 2021-09-02

- TLS (`tls`)
  - Extensions (`extensions`)
    - add [application-layer protocol negotiation](https://www.rfc-editor.org/rfc/rfc5077.html) extension related
      messages (\#40)
    - add [encrypt-then-MAC](https://www.rfc-editor.org/rfc/rfc7366.html) extension related messages (\#40)
    - add [extended master secret](https://www.rfc-editor.org/rfc/rfc7627.html) extension related messages (\#40)
    - add [next protocol negotiation](https://tools.ietf.org/id/draft-agl-tls-nextprotoneg-03.html) extension related
      messages (\#40)
    - add [renegotiation indication](https://www.rfc-editor.org/rfc/rfc5746.html) extension related messages (\#40)
    - add [session ticket](https://www.rfc-editor.org/rfc/rfc5077.html) extension related messages (\#40)

## 0.6.0 - 2021-05-27

### Features

- HTTP(S) (`http`)
  - Headers (`headers`)
    - supports header wire format parsing
    - add parsers for generic headers
      ([Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type),
      [Server](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server))
    - add parsers for cache related headers ([Age](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Age),
      [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control),
      [Date](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Date),
      [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag),
      [Expires](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expires),
      [Last-Modified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified),
      [Pragma](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Pragma))
    - add parsers for security related headers
      ([Expect-CT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expect-CT),
      [Expect-Staple](https://scotthelme.co.uk/designing-a-new-security-header-expect-staple),
      [Referrer-Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy),
      [Strict-Transport-Security](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security),
      [X-Content-Type-Options](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options),
      [X-Frame-Options](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options))
- TLS (`tls`)
  - Versions (`versions`)
    - add [protocol version 1.3](https://tools.ietf.org/html/rfc8446) related messages (\#20)
  - Cipher Suites (`ciphers`)
    - add [cipher suites](https://tools.ietf.org/html/rfc8446#appendix-B.4) relate to version 1.3 (\#20)
  - Diffie-Hellman (`dhparams`)
    - add [supported groups](https://tools.ietf.org/html/rfc8446#section-4.2.7) relate to version 1.3 (\#20)
  - Elliptic Curves (`curves`)
    - add [supported groups](https://tools.ietf.org/html/rfc8446#section-4.2.7) relate to version 1.3 (\#20)
  - Signature Algorithms (`sigalgos`)
    - add [signature algorithms](https://tools.ietf.org/html/rfc8446#section-4.2.3) relate to version 1.3 (\#20)

## 0.5.0 - 2021-04-08

### Features

- Generic
  - add parser for [text-based protocols](https://en.wikipedia.org/wiki/Text-based_protocol) (\#21)
- SSH (`ssh`)
  - Versions (`versions`)
    - add [protocol version exchange](https://tools.ietf.org/html/rfc4253#section-4.2) related messages (\#21)
- SSH 2.0 (`ssh2`)
  - Cipher Suites (`ciphers`)
    - add [algorithm negotiation](https://tools.ietf.org/html/rfc4253#section-7.1) related messages (\#21)

### Usability

- Generic
  - show attributes in user-friendly order in Markdown output (\#30)
  - use human readable algorithms names in Markdown output (\#32)
  - add human readable descriptions for exceptions (\#33)

## 0.4.0 - 2021-01-30

### Features

- TLS (`tls`)
  - Generic
    - add [LDAP](https://en.wikipedia.org/wiki/Lightweight_Directory_Access_Protocol) related messages (\#23)
  - Client Public Key Request (`pubkeyreq`)
    - add [client public key request](https://tools.ietf.org/html/rfc2246#section-7.4.4) related messages (\#24)

### Improvements

- Generic
  - add [OID](https://en.wikipedia.org/wiki/Object_identifier) to algorithms

## 0.3.1 - 2020-09-15

### Features

- Generic
  - [Markdown](https://en.wikipedia.org/wiki/Markdown) serializable format (\#19)

### Improvements

- TLS (`tls`)
  - Cipher Suites (`ciphers`)
    - add missing `ECDHE_PSK` cipher suites (\#7)
    - add [GOST](https://en.wikipedia.org/wiki/GOST) cipher suites
    - add missing draft ECC cipher suites (\#9)
    - add missing [FIPS](https://en.wikipedia.org/wiki/FIPS_140-2) cipher suites (\#11)
    - add [CECPQ1](https://en.wikipedia.org/wiki/CECPQ1) cipher suites (\#12)

cryptoparser/tls/mysql.py
# -*- coding: utf-8 -*-

import abc
import enum
import six

import attr

from cryptoparser.common.base import Serializable, OneByteEnumComposer, OneByteEnumParsable
from cryptoparser.common.parse import ByteOrder, ComposerBinary, ParsableBase, ParserBinary
from cryptoparser.common.exception import NotEnoughData, InvalidValue


class MySQLVersion(enum.IntEnum):
    MYSQL_9 = 0x09
    MYSQL_10 = 0x0a


class MySQLCapability(enum.IntEnum):
    CLIENT_LONG_PASSWORD = 0x00000001
    CLIENT_FOUND_ROWS = 0x00000002
    CLIENT_LONG_FLAG = 0x00000004
    CLIENT_CONNECT_WITH_DB = 0x00000008
    CLIENT_NO_SCHEMA = 0x00000010
    CLIENT_COMPRESS = 0x00000020
    CLIENT_ODBC = 0x00000040
    CLIENT_LOCAL_FILES = 0x00000080
    CLIENT_IGNORE_SPACE = 0x00000100
    CLIENT_PROTOCOL_41 = 0x00000200
    CLIENT_INTERACTIVE = 0x00000400
    CLIENT_SSL = 0x00000800
    CLIENT_IGNORE_SIGPIPE = 0x00001000
    CLIENT_TRANSACTIONS = 0x00002000
    CLIENT_RESERVED = 0x00004000
    CLIENT_SECURE_CONNECTION = 0x00008000
    CLIENT_MULTI_STATEMENTS = 0x00010000
    CLIENT_MULTI_RESULTS = 0x00020000
    CLIENT_PS_MULTI_RESULTS = 0x00040000
    CLIENT_PLUGIN_AUTH = 0x00080000
    CLIENT_CONNECT_ATTRS = 0x00100000
    CLIENT_PLUGIN_AUTH_LENENC_CLIENT_DATA = 0x00200000
    CLIENT_CAN_HANDLE_EXPIRED_PASSWORDS = 0x00400000
    CLIENT_SESSION_TRACK = 0x00800000
    CLIENT_DEPRECATE_EOF = 0x01000000


class MySQLStatusFlag(enum.IntEnum):
    SERVER_STATUS_IN_TRANS = 0x0001
    SERVER_STATUS_AUTOCOMMIT = 0x0002
    SERVER_MORE_RESULTS_EXISTS = 0x0008
    SERVER_STATUS_NO_GOOD_INDEX_USED = 0x0010
    SERVER_STATUS_NO_INDEX_USED = 0x0020
    SERVER_STATUS_CURSOR_EXISTS = 0x0040
    SERVER_STATUS_LAST_ROW_SENT = 0x0080
    SERVER_STATUS_DB_DROPPED = 0x0100
    SERVER_STATUS_NO_BACKSLASH_ESCAPES = 0x0200
    SERVER_STATUS_METADATA_CHANGED = 0x0400
    SERVER_QUERY_WAS_SLOW = 0x0800
    SERVER_PS_OUT_PARAMS = 0x1000
    SERVER_STATUS_IN_TRANS_READONLY = 0x2000
    SERVER_SESSION_STATE_CHANGED = 0x4000


@attr.s(frozen=True)
class MySQLCharacterSetParams(Serializable):
    code = attr.ib(validator=attr.validators.instance_of(int))
    name = attr.ib(validator=attr.validators.instance_of(six.string_types))
    collate_name = attr.ib(validator=attr.validators.instance_of(six.string_types))


class MySQLCharacterSetFactory(OneByteEnumParsable):
    @classmethod
    def get_enum_class(cls):
        return MySQLCharacterSet

    @abc.abstractmethod
    def compose(self):
        raise NotImplementedError()


class MySQLCharacterSet(OneByteEnumComposer, enum.Enum):
    BIG5 = MySQLCharacterSetParams(
        code=1,
        name='big5',
        collate_name='big5_chinese_ci',
    )
    DEC8 = MySQLCharacterSetParams(
        code=3,
        name='dec8',
        collate_name='dec8_swedish_ci',
    )
    CP850 = MySQLCharacterSetParams(
        code=4,
        name='cp850',
        collate_name='cp850_general_ci',
    )
    HP8 = MySQLCharacterSetParams(
        code=6,
        name='hp8',
        collate_name='hp8_english_ci',
    )
    KOI8R = MySQLCharacterSetParams(
        code=7,
        name='koi8r',
        collate_name='koi8r_general_ci',
    )
    LATIN1 = MySQLCharacterSetParams(
        code=8,
        name='latin1',
        collate_name='latin1_swedish_ci',
    )
    LATIN2 = MySQLCharacterSetParams(
        code=9,
        name='latin2',
        collate_name='latin2_general_ci',
    )
    SWE7 = MySQLCharacterSetParams(
        code=10,
        name='swe7',
        collate_name='swe7_swedish_ci',
    )
    ASCII = MySQLCharacterSetParams(
        code=11,
        name='ascii',
        collate_name='ascii_general_ci',
    )
    UJIS = MySQLCharacterSetParams(
        code=12,
        name='ujis',
        collate_name='ujis_japanese_ci',
    )
    SJIS = MySQLCharacterSetParams(
        code=13,
        name='sjis',
        collate_name='sjis_japanese_ci',
    )
    HEBREW = MySQLCharacterSetParams(
        code=16,
        name='hebrew',
        collate_name='hebrew_general_ci',
    )
    TIS620 = MySQLCharacterSetParams(
        code=18,
        name='tis620',
        collate_name='tis620_thai_ci',
    )
    EUCKR = MySQLCharacterSetParams(
        code=19,
        name='euckr',
        collate_name='euckr_korean_ci',
    )
    KOI8U = MySQLCharacterSetParams(
        code=22,
        name='koi8u',
        collate_name='koi8u_general_ci',
    )
    GB2312 = MySQLCharacterSetParams(
        code=24,
        name='gb2312',
        collate_name='gb2312_chinese_ci',
    )
    GREEK = MySQLCharacterSetParams(
        code=25,
        name='greek',
        collate_name='greek_general_ci',
    )
    CP1250 = MySQLCharacterSetParams(
        code=26,
        name='cp1250',
        collate_name='cp1250_general_ci',
    )
    GBK = MySQLCharacterSetParams(
        code=28,
        name='gbk',
        collate_name='gbk_chinese_ci',
    )
    LATIN5 = MySQLCharacterSetParams(
        code=30,
        name='latin5',
        collate_name='latin5_turkish_ci',
    )
    ARMSCII8 = MySQLCharacterSetParams(
        code=32,
        name='armscii8',
        collate_name='armscii8_general_ci',
    )
    UTF8 = MySQLCharacterSetParams(
        code=33,
        name='utf8',
        collate_name='utf8_general_ci',
    )
    UCS2 = MySQLCharacterSetParams(
        code=35,
        name='ucs2',
        collate_name='ucs2_general_ci',
    )
    CP866 = MySQLCharacterSetParams(
        code=36,
        name='cp866',
        collate_name='cp866_general_ci',
    )
    KEYBCS2 = MySQLCharacterSetParams(
        code=37,
        name='keybcs2',
        collate_name='keybcs2_general_ci',
    )
    MACCE = MySQLCharacterSetParams(
        code=38,
        name='macce',
        collate_name='macce_general_ci',
    )
    MACROMAN = MySQLCharacterSetParams(
        code=39,
        name='macroman',
        collate_name='macroman_general_ci',
    )
    CP852 = MySQLCharacterSetParams(
        code=40,
        name='cp852',
        collate_name='cp852_general_ci',
    )
    LATIN7 = MySQLCharacterSetParams(
        code=41,
        name='latin7',
        collate_name='latin7_general_ci',
    )
    CP1251 = MySQLCharacterSetParams(
        code=51,
        name='cp1251',
        collate_name='cp1251_general_ci',
    )
    UTF16 = MySQLCharacterSetParams(
        code=54,
        name='utf16',
        collate_name='utf16_general_ci',
    )
    UTF16LE = MySQLCharacterSetParams(
        code=56,
        name='utf16le',
        collate_name='utf16le_general_ci',
    )
    CP1256 = MySQLCharacterSetParams(
        code=57,
        name='cp1256',
        collate_name='cp1256_general_ci',
    )
    CP1257 = MySQLCharacterSetParams(
        code=59,
        name='cp1257',
        collate_name='cp1257_general_ci',
    )
    UTF32 = MySQLCharacterSetParams(
        code=60,
        name='utf32',
        collate_name='utf32_general_ci',
    )
    BINARY = MySQLCharacterSetParams(
        code=63,
        name='binary',
        collate_name='binary',
    )
    GEOSTD8 = MySQLCharacterSetParams(
        code=92,
        name='geostd8',
        collate_name='geostd8_general_ci',
    )
    CP932 = MySQLCharacterSetParams(
        code=95,
        name='cp932',
        collate_name='cp932_japanese_ci',
    )
    EUCJPMS = MySQLCharacterSetParams(
        code=97,
        name='eucjpms',
        collate_name='eucjpms_japanese_ci',
    )
    GB18030 = MySQLCharacterSetParams(
        code=248,
        name='gb18030',
        collate_name='gb18030_chinese_ci',
    )
    UTF8MB4 = MySQLCharacterSetParams(
        code=255,
        name='utf8mb4',
        collate_name='utf8mb4_0900_ai_ci',
    )


class MySQLPacketBase(ParsableBase):
    @classmethod
    @abc.abstractmethod
    def _parse(cls, parsable):
        raise NotImplementedError()

    @abc.abstractmethod
    def compose(self):
        raise NotImplementedError()


@attr.s
class MySQLRecord(ParsableBase):
    HEADER_SIZE = 4

    packet_number = attr.ib(validator=attr.validators.instance_of(six.integer_types))
    packet_bytes = attr.ib(validator=attr.validators.instance_of((bytes, bytearray)))

    @classmethod
    def _parse(cls, parsable):
        if len(parsable) < cls.HEADER_SIZE:
            raise NotEnoughData(cls.HEADER_SIZE - len(parsable))

        parser = ParserBinary(parsable, byte_order=ByteOrder.LITTLE_ENDIAN)

        parser.parse_numeric('packet_length', 3)
        parser.parse_numeric('packet_number', 1)
        parser.parse_raw('packet_bytes', parser['packet_length'])

        return MySQLRecord(
            packet_number=parser['packet_number'],
            packet_bytes=parser['packet_bytes'],
        ), parser.parsed_length

    def compose(self):
        composer = ComposerBinary(byte_order=ByteOrder.LITTLE_ENDIAN)

        composer.compose_numeric(len(self.packet_bytes), 3)
        composer.compose_numeric(self.packet_number, 1)
        composer.compose_raw(self.packet_bytes)

        return composer.composed_bytes


@attr.s
class MySQLHandshakeV10(MySQLPacketBase):  # pylint: disable=too-many-instance-attributes
    protocol_version = attr.ib(validator=attr.validators.in_(MySQLVersion))
    server_version = attr.ib(validator=attr.validators.instance_of(six.string_types))
    connection_id = attr.ib(validator=attr.validators.instance_of(six.integer_types))
    auth_plugin_data = attr.ib(validator=attr.validators.instance_of((bytes, bytearray)))
    capabilities = attr.ib(validator=attr.validators.deep_iterable(
        member_validator=attr.validators.instance_of(MySQLCapability),
    ))
    character_set = attr.ib(
        default=MySQLCharacterSet.UTF8, validator=attr.validators.optional(attr.validators.in_(MySQLCharacterSet))
    )
    states = attr.ib(default={}, validator=attr.validators.deep_iterable(
        member_validator=attr.validators.instance_of(MySQLStatusFlag),
    ))
    auth_plugin_data_2 = attr.ib(
        default=None,
        validator=attr.validators.optional(attr.validators.instance_of((bytes, bytearray))),
    )
    auth_plugin_name = attr.ib(
        default=None,
        validator=attr.validators.optional(attr.validators.instance_of(six.string_types))
    )

    MINIMUM_SIZE = 33

    @classmethod
    def _parse(cls, parsable):
        if len(parsable) < cls.MINIMUM_SIZE:
            raise NotEnoughData(cls.MINIMUM_SIZE - len(parsable))

        parser = ParserBinary(parsable, byte_order=ByteOrder.LITTLE_ENDIAN)

        parser.parse_numeric('protocol_version', 1, MySQLVersion)
        parser.parse_string_null_terminated('server_version', 'ascii')
        parser.parse_numeric('connection_id', 4)
        parser.parse_raw('auth_plugin_data', 8)
        parser.parse_raw('filler', 1)
        del parser['filler']
        parser.parse_numeric_flags('capabilities', 2, MySQLCapability)
        parser.parse_parsable('character_set', MySQLCharacterSetFactory)
        parser.parse_numeric_flags('states', 2, MySQLStatusFlag)
        parser.parse_numeric_flags('capabilities_2', 2, MySQLCapability, shift_left=16)
        capabilities = set(parser['capabilities']) | set(parser['capabilities_2'])
        del parser['capabilities_2']

        parser.parse_numeric('auth_plugin_data_len', 1)
        auth_plugin_data_len = parser['auth_plugin_data_len']
        del parser['auth_plugin_data_len']

        parser.parse_raw('reserved', 10)
        del parser['reserved']

        if MySQLCapability.CLIENT_PLUGIN_AUTH in capabilities:
            if not auth_plugin_data_len:
                raise InvalidValue(auth_plugin_data_len, cls, 'auth_plugin_data_len')

            auth_plugin_data_2_len = auth_plugin_data_len - 8
            parser.parse_raw('auth_plugin_data_2', auth_plugin_data_2_len)

        if MySQLCapability.CLIENT_PLUGIN_AUTH in capabilities:
            parser.parse_string_null_terminated('auth_plugin_name', 'ascii')

        params = dict(parser)
        params['capabilities'] = capabilities

        return cls(**params), parser.parsed_length

    def compose(self):
        composer = ComposerBinary(byte_order=ByteOrder.LITTLE_ENDIAN)

        composer.compose_numeric(self.protocol_version, 1)
        composer.compose_string_null_terminated(self.server_version, 'ascii')
        composer.compose_numeric(self.connection_id, 4)

        composer.compose_raw(self.auth_plugin_data)

        composer.compose_raw(b'\x00')  # filler

        capabilities = [capability for capability in self.capabilities if capability.value < 2 ** 16]
        composer.compose_numeric_flags(capabilities, 2)

        capabilities_2 = [capability for capability in self.capabilities if capability.value >= 2 ** 16]
        composer.compose_parsable(self.character_set)
        composer.compose_numeric_flags(self.states, 2)
        composer.compose_numeric_flags(capabilities_2, 2, shift_right=16)

        if MySQLCapability.CLIENT_PLUGIN_AUTH in self.capabilities:
            auth_plugin_data_len = 8
            if self.auth_plugin_data_2:
                auth_plugin_data_len += len(self.auth_plugin_data_2)

            composer.compose_numeric(auth_plugin_data_len, 1)
        else:
            composer.compose_numeric(0, 1)

        composer.compose_raw(10 * b'\x00')  # reserved
        if self.auth_plugin_data_2:
            composer.compose_raw(self.auth_plugin_data_2)

        if MySQLCapability.CLIENT_PLUGIN_AUTH in self.capabilities:
            composer.compose_string_null_terminated(self.auth_plugin_name, 'ascii')

        return composer.composed_bytes


@attr.s
class MySQLHandshakeSslRequest(MySQLPacketBase):
    capabilities = attr.ib(validator=attr.validators.deep_iterable(
        member_validator=attr.validators.instance_of(MySQLCapability),
    ))
    max_packet_size = attr.ib(default=0xffff, validator=attr.validators.instance_of(six.integer_types))
    character_set = attr.ib(default=None, validator=attr.validators.optional(attr.validators.in_(MySQLCharacterSet)))

    MINIMUM_SIZE = 5

    def __attrs_post_init__(self):
        if MySQLCapability.CLIENT_PROTOCOL_41 in self.capabilities:
            if self.character_set is None:
                self.character_set = MySQLCharacterSet.UTF8
        else:
            if self.max_packet_size >= 2 ** 24:
                raise ValueError(self.max_packet_size)
            for capability in self.capabilities:
                if capability.value >= 2 ** 16:
                    raise ValueError(self.max_packet_size)

    @classmethod
    def _parse(cls, parsable):
        if len(parsable) < cls.MINIMUM_SIZE:
            raise NotEnoughData(cls.MINIMUM_SIZE - len(parsable))

        parser = ParserBinary(parsable, byte_order=ByteOrder.LITTLE_ENDIAN)

        parser.parse_numeric_flags('capabilities', 2, MySQLCapability)
        if MySQLCapability.CLIENT_PROTOCOL_41 in parser['capabilities']:
            parser.parse_numeric_flags('capabilities_2', 2, MySQLCapability, shift_left=16)
            parser.parse_numeric('max_packet_size', 4)
            parser.parse_parsable('character_set', MySQLCharacterSetFactory)
            parser.parse_raw('reserved', 23)
            del parser['reserved']

            character_set = parser['character_set']
            capabilities = parser['capabilities'] | parser['capabilities_2']
        else:
            parser.parse_numeric('max_packet_size', 3)
            capabilities = parser['capabilities']
            character_set = None

        return cls(capabilities, parser['max_packet_size'], character_set), parser.parsed_length

    def compose(self):
        composer = ComposerBinary(byte_order=ByteOrder.LITTLE_ENDIAN)

        if MySQLCapability.CLIENT_PROTOCOL_41 in self.capabilities:
            composer.compose_numeric_flags(self.capabilities, 4)
            composer.compose_numeric(self.max_packet_size, 4)
            composer.compose_parsable(self.character_set)
            composer.compose_raw(23 * b'\x00')
        else:
            composer.compose_numeric_flags(self.capabilities, 2)
            composer.compose_numeric(self.max_packet_size, 3)

        return composer.composed_bytes
